### PR TITLE
Fixing copy and cast from exile effects

### PR DESCRIPTION
**Pull request by alexander-novo**
Details in #10435

### CODE REVIEW

**Comment by alexander-novo** [comment]
Debugging this test shows that delayed blast fireball is getting cast from outside the game rather than from exile, as should be the case:

image

**Comment by Grath** [comment]
Just as a warning, from experience with poking at this sort of effect: You may run into some broken nonsense around copying kicked permanents and Zone Change Counters.

**Comment by JayDi85** [comment]
`addCard` is fine -- it must be outside. I think `Player->cast` can be modified with extra param, so copy abilitity can override `formZone`.
shot_230607_005543
shot_230607_005830

BUT it must be good tested. There are some code like `Card->cast` with card moving code inside (see `maybeRemoveFromSourceZone`), it can be potentially broken. So unit test must check copied card zone after cast -- is it moved to battlefield after cast -- copied card, not a permanent). ZCC can be checked too somehow (possible use case: copy card with the kicker ability).

Also possible bad use cases to check (I don't know is it possible in real game):
- somehow copy a battlefield's permanent as a card (fromZone = battlefield);
- somehow copy stack's spell as a card (fromZone = stack)

**Comment by JayDi85** [comment]
Well, if you modify `copyCard` and `addCard` to support starting zone setup with strict check (disallow battlefield zone) then it must works fine (it's a best fix). `checkStateBasedActions` removes all copied cards to outside all the time, but it's ok:
shot_230607_013233

**Comment by alexander-novo** [comment]
Perfect, thanks for the reply. I'll look into that then. Also gonna add a couple more tests for some edge cases I'm worried about, such as [[Prosper, Tome-Bound]] - which shouldn't work with Mizzix's Mastery style effects. It currently doesn't but I'm worried it's because of the incorrect zone rather than properly paying attention to whether or not the spell cast comes from a card

**Comment by alexander-novo** [comment]
Okay, I changed `copyCard` to create the copy in the zone that the original is in. It turns out there were some effects that already did this manually ([[Isochron Scepter]] and [[God-Eternal Kefnet]]), but it seemed more appropriate to just have the function do it automatically, since it's only used for these kinds of effects which need to be done properly (and often weren't). Those cards were changed to no longer manually move the card to the correct zone (for some reason this was causing state-based effects to fail to remove the copies).

Also, as I feared, Prosper and a couple of other cards that care about playing *cards* from a specific zone no longer work properly, because they never checked to make sure what was being cast was actually a card. I actually reported that as an issue earlier (#9931). It looks like that was only manifesting with some effects (such as Mnemonic Deluge) because roughly 1/3 of these copy then cast from another zone effects were setup to properly move the copied card before casting it.

**Comment by alexander-novo** [comment]
Okay, I wrote a new common triggered ability for this "plays a card" wording. I think I got every card that uses this wording, but I'm not 100% sure. Notably [[Prowl, Stoic Strategist]] will need to use this new methodology, but it hasn't been implemented yet.

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage.Tests/src/test/java/org/mage/test/cards/single/clb/DelayedBlastFireballTest.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public class DelayedBlastFireballTest extends CardTestPlayerBase {
+    static final String fireball = "Delayed Blast Fireball";
+
+    @Test
+    // Author: alexander-novo
+    // Issue: magefree/mage#10435
+    // If you create a copy of a card in exile and then cast that copy, it should be cast from exile.
+    // But if we do this with Delayed Blast Fireball (and Mizzix's Mastery, for instance), it won't deal the extra damage for casting from exile.
+    public void testCopyCardAndCastFromExile() {
+        String mastery = "Mizzix's Mastery";
```
[the remark this answers is not in the document]
Comment (alexander-novo) [reply Mage.Tests/src/test/java/org/mage/test/cards/single/clb/DelayedBlastFireballTest.java:18]: Oop. It looks like the overload is probably broken. I'll add a test case for this.

File: Mage/src/main/java/mage/abilities/common/PlayCardTriggeredAbility.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import mage.abilities.effects.Effect;
+import mage.constants.TargetController;
+import mage.constants.Zone;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+
+// Author: alexander-novo
+// A triggered ability for cards which say "whenever <someone> play(s) a card..."
+public class PlayCardTriggeredAbility extends TriggeredAbilityImpl {
+
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/common/PlayCardTriggeredAbility.java:14]: Can also be used for [[Juju Bubble]]

Comment (alexander-novo) [reply Mage/src/main/java/mage/abilities/common/PlayCardTriggeredAbility.java:14]: Cool thanks. I had problems finding cards that needed to use this.

**Comment by alexander-novo** [comment]
I got the 16/1000 =(

**Comment by JayDi85** [comment]
@alexander-novo is PR fixed #9931 too?

> Also, as I feared, Prosper and a couple of other cards that care about playing cards from a specific zone no longer work properly, because they never checked to make sure what was being cast was actually a card

**Comment by alexander-novo** [comment]
@JayDi85 yup
